src/repr/src/row/encoding.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A permanent storage encoding for rows.
+//!
+//! To minimize cycles spent encoding/decoding, we use Row's internal storage
+//! format, with a prefix to allow for migrations if the internal format
+//! changes.
+//!
+//! ```none
+//! <version: 0u8><v0_encoding: &[u8]>
+//! - Currently the only value for `version` is `0u8`.
+//!
+//! <v0_encoding> is <len_bytes: u64><row_data: &[u8]>
+//! - `len_bytes` is little endian.
+//! - `row_data` is `len_bytes` long.
+//!
+//! <row_data> is repeated <tag>(<datum: &[u8]>)?
+//! - `tag` is one of the following values and determines the presence, length,
+//!   and encoding of `datum`.
+//! - All `datum` integers, signed and unsigned, are little endian.
+//!
+//! <tag=0u8> is Datum::Null
+//! <tag=1u8> is Datum::False
+//! <tag=2u8> is Datum::True
+//! <tag=3u8><val: i32> is Datum::I32(val)
+//! WIP finish this
```

Comment:
Thoughts on how I should write this out? Maybe something like https://developers.google.com/protocol-buffers/docs/reference/proto3-spec?